### ISSUE TITLE
Prove out a good solution for the default route use case

### ISSUE THREAD

**Issue by rictic** [PolymerElements/app-route#68]
*As a user of carbon-route, I want to be able to react when none of my carbon-routes in a section of code match the current path. For example, I may want to display a 404 page, or redirect to my home page.*

One solution may be adding the concept of being given a selection that doesn't match to `<iron-pages>`.

**Comment by cdata** [comment]
The way I have worked around this until now is measuring for changes to parameterized values and then checking to see if the new value isn't among the expected ones. This is kind of clunky, and it would be nice to have a more idiomatic approach.

**Comment by rictic** [comment]
Yeah, ideally we wouldn't require repeating the options in both the HTML and a JS observer.

**Comment by TimvdLippe** [comment]
I could come up with two possible solutions here:

1. The implementation is included in `carbon-route`. For example some sort of meta-element that keeps track of sub-routes and set the route to `*` if none route was matching.
* All receiving [`IronSelectableBehavior`](https://elements.polymer-project.org/elements/iron-selector?active=Polymer.IronSelectableBehavior) must handle a case where a route does not match.

The second solution would be probably troublesome for custom implementations of `IronSelectableBehavior`. The first solution however requires a new meta-element, which would be somewhat undesirable. WDYT?

**Comment by kenchris** [comment]
I would also like this a lot, like my app only supports /#!/home /#!/list etc. I was trying to filter like

```
        properties: {
          route: {
            type: Object,
            notify: true,
            observer: "onRouteChanged"
          }
        },

        onRouteChanged: function() {
          console.log(this.route);
          if (["home", "list"].indexOf(this.route.page) < 0) {
            console.log("I don't understand this page");
            window.location.hash = '!/home';
          }
        },
```

But for some reason the location.hash isn't updated. When I update that manually elsewhere, it works fine:

```
        gotoHome: function() {
          window.location.hash = '!/home';
        },
```

It might be nice to be able to give an array of accepted values (maybe accepting something like gulp-glob does) and if not matching having that redirect to an existing route or to a 404 page

**Comment by TimvdLippe** [comment]
A proposal has been submitted for [`IronSelectableBehavior`](https://github.com/PolymerElements/iron-selector/issues/116) to implement the earlier stated second solution.

**Comment by kenchris** [comment]
That is a solution which I can support, but I don't think it might be enough in all cases, like it will mean that I will have to do the fallback at all the source spots which - depending on the project - might not be the best solution.

Having both options would be better.

**Comment by kenchris** [comment]
What about a kind of redirector?

```
    <carbon-location route="{{route}}" use-hash-as-path></carbon-location>
    <carbon-redirect route="{{route}}" data="{{validRoute}}" match="/^!\/(?!(home|list)$)/" redirect="!/home">
    <carbon-route
      route="{{validRoute}}"
      pattern="!/:page"
      data="{{routeData}}"
      tail="{{routeTail}}">
    </carbon-route>
```

It could of course filter using a method instead

**Comment by TimvdLippe** [comment]
@kenchris That would however mean that the `carbon-redirect` needs to know every possible subpath. This is in my eyes the responsibility of a selector which based on the input selects an element/option.

I am not sure what you mean with "the fallback at all the source spots". The `carbon-route` does not decide what the fallback is, this is up to the receiving element (for example an `iron-pages`) to decide how to process the selected value.

To keep the separation of concerns, I am still convinced that the implementation details for the selector element.

**Comment by kenchris** [comment]
I mean that the route might control a lot more in my app than decide what iron-page is shown, so it means that I need to handle the routes at each of these spots in my code where I make such decisions - which might not be optimal solution, which is why I like being able to handle redirection in one place.

The ```carbon-redirect``` doesn't need to handle every possible subpath, but at least filter out some of it, meaning that I know exactly what I might need to handle elsewhere

**Comment by abdonrd** [comment]
I haven't yet started testing the `carbon-route` a lot, but I like the @TimvdLippe option; delegate that responsibility in the selector element: `iron-lazy-pages`?.

**Comment by ktiedt** [comment]
A valid point was raised though, this forces every selector to be aware of
a "default" where as if it was a feature of the router element it could
handle it in one place which is also a very valid design point. Technically
the router should handle the selection value and the selector the display
aspect.

**Comment by olegomon** [comment]
I agree with @kedit Please consider also conditional redirects. If the user is logged in I would like to redirect to a default app route like /app or to /home otherwise.

**Comment by rictic** [comment]
Agreed that we should have super solid redirection support, both for 404-style use cases as well as not-logged-in and others. An element like `carbon-redirect` as sketched out by @kenchris should be fairly straightforwards to write after #44 is fixed.

@ktiedt With Tim's proposal the handling of the default is done in a single place. For me it's just a matter of which is more elegant. The element that's handling switching between pages already __must__ know the list of pages that it can display. We could duplicate that knowledge in the routing system, but I'm not convinced that's a win.

With Tim's design it's super DRY:

```html
<carbon-route route="{{route}}" pattern="/:category" data="{{data}}">
</carbon-route>

<iron-pages selected="{{data.category}}" attr-for-selected="name" default-selected="404">
  <home-page name="home"></home-page>
  <user-page name="user"></user-page>
  <404-page name="404"></404-page>  
</iron-pages>
```

**Comment by ktiedt** [comment]
@rictic it is a single place in most circumstances but ultimately it is not always a single point.

What happens when you have a nested route like /Users and /Users/ID

Users/ID may be a nested iron-pages to toggle between "generic user page" vs "user profile page for ID" Now 2 place potentially need "default" values.

The one problem I see with the pages element handling things (after reading the other comments here) is that the pages element is that it now is taking on a whole new role. It is no longer a responsive element that shows you what you ask for, but a reactive element that shows you what you werent looking for and instead of being the "follower" is now the "leader" and must communicate that state back UP to the router.

**Comment by TimvdLippe** [comment]
@ktiedt In that situation, we essentially have a declarative form of the previous `page.js` implementation. I do not see a harm in creating a separate element to provide that functionality, but I do think that `carbon-route` is not the place to implement it.

[In the gist I created](https://gist.github.com/TimvdLippe/9bfebc20aec75dd36126b8162c310dff#file-user-page-html-L3) for the default selector, you can see how the `users/id` case is handled. In this case, it is the responsibility of the `user-page` to determine what it should show. This not only makes the element reusable, it also separates the responsibilities for each page on its own.

The difference in solutions is a fundamental one and I can see why others would prefer a monolithic `page.js`-like solution. Personally the philosophy adopted with `carbon-route` fundamentally changed the way I think about routing and to me seems like a very big win. If there is a large group of developers who would like the other solution, why don't we create a separate element for it?

**Comment by ktiedt** [comment]
I still see a functional shortcoming that appears to be overlooked in this case. Which is why leaving the routing to the router (not the page) makes more sense.

Without an working demo to play with here, let me go through some logic steps, maybe I am missing something and you can point out where I'm not seeing this working as expected.

**Current state of carbon-router**
1. User loads `mysite.com/non-existent-page`
2. router doesn't know what to do

**proposed selector changes**
1. User loads `mysite.com/non-existent-page`
2. router doesn't know what to do
3. selector says "you must want this" and displays default page
4. router still has no idea what changed (selector is bound to `data` which doesnt correlate to a `route`
5. display is now out of sync with what was requested by the URL

**proposal of allowing router to handle 404s**
1. User loads `mysite.com/non-existent-page`
2. router kicks off the redirect
3. router changes URL to redirect
4. selector selects the page associated with 404
5. display is still in sync with what was requested by the URL

**Comment by rictic** [comment]
@ktiedt As far as I can tell, the only user-visible difference between the selector changes and your proposal is that the  `/non-existant-page` redirects to `/my-404-page`. IMO some people will want to redirect when there's a 404 and some won't, and we should support both. The selector change should give you the hooks to redirect if you want.

With regard to whether a given piece of information is part of 'routing' or should be known to 'the router' – I've come to think that these categories haven't been drawn very well in the past and that it's worth taking a very skeptical eye towards them. Sorry if I'm belaboring that point to you Karl, I know we chatted about this at the meetup, but I just wanted to mention it here to give context for why we're taking such a deliberative tack with carbon-route.

**Comment by ktiedt** [comment]
No problem, my only concern is that, effective route changes should not happen w/o the knowledge of the router and without updating the route state. That seems like a contract that should not be easily broken or that should require the user to go out of there way to achieve.

**Comment by rictic** [comment]
Yeah, that's reasonable. I can see the value in both ideals. We'll take a stab at the iron-selector changes, as they are small and reasonable in isolation of this debate, and we can see how well it works for a variety of use cases.

**Comment by TimvdLippe** [comment]
I just pushed an update to the `iron-selector` PR. This allows the developer to use the data-binding system to let the router be notified of changes.

Example:
```html
<iron-lazy-pages selected="{{routeData.page}}" attr-for-selected="data-route" default-selected="404">
  <template is="iron-lazy-page" data-route="">
    Home
   <a href="/store">Visit the store.</a>
  </template>
  <template is="iron-lazy-page" data-route="404">
    Oops you hit a 404.
    <a href="/">Head back home.</a>
  </template>
</iron-lazy-pages>
```

* When you open `domain.com/asdf` the selector will change to the 404 and the url is updated to `domain.com/404`.
* However if you use one-way-binding (`selected="[[routeData.page]]"`) you still see the 404 message, but your url is not updated.

Therefore it is up to developer to decide whether he/she wants to force the url update or not.

**Comment by ktiedt** [comment]
@TimvdLippe can you apply this example here to a nested route? Myself and my teammate are not seeing a solution for this in these changes, and it is still something we are going to need to support in our projects...

**Comment by TimvdLippe** [comment]
As per slack conversation: Yes you can apply this to any nested route, but it will not effect any global route. As you can see [here](https://github.com/TimvdLippe/polymer-lazy-application/blob/master/app/store/store-page.html#L51-L53) if you open `localhost:3636/store/invalidstore` it will redirect you to `localhost:3636/store/not-found`

**Comment by TimvdLippe** [comment]
The property `fallback-selection` on `iron-selector`s has been available for quite some time now. Has it proved to be successful in the provided use-cases?

**Comment by JosefJezek** [comment]
@TimvdLippe How to handle invalid subroute? Example route: `localhost:3636/store/amsterdam/invalidpage`

**Comment by TimvdLippe** [comment]
@JosefJezek Add a `fallback-selection="invalid-category"` on the `paper-listbox` and add `<paper-item data-category="invalid-category">Invalid category</paper-item>`

**Comment by rictic** [comment]
I think it's safe to close this issue, as fallback-selection has proven to be a quite useful solution.

**Comment by chrismbeckett** [comment]
I am using the <app-route> element (not <carbon-route> as noted above), but fallback-selection is only working on the top-level app element. If I have a nested <app-route> in a sub-page, the fallback-selection is working, but it is not updating the URL.
